robot/packages/cloud_package_manger.go

```diff
@@ -0,0 +1,550 @@
+	packagesDataDir := filepath.Join(packagesDir, ".data")
+
+	if err := os.MkdirAll(packagesDir, 0o700); err != nil {
+		return nil, err
+	}
+
+	if err := os.MkdirAll(packagesDataDir, 0o700); err != nil {
+		return nil, err
+	}
+
+	return &cloudManager{
+		client:          client,
+		httpClient:      http.Client{Timeout: time.Minute * 30},
+		packagesDir:     packagesDir,
+		packagesDataDir: packagesDataDir,
+		managedPackages: make(map[PackageName]*managedPackage),
+		logger:          logger.Named("package_manager"),
+	}, nil
+}
+
+// PackagePath returns the package if it exists and already download. If it does not exist it returns a ErrPackageMissing error.
+//
+// TODO(adam): test.
+func (m *cloudManager) PackagePath(name PackageName) (string, error) {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	p, ok := m.managedPackages[name]
+	if !ok {
+		return "", ErrPackageMissing
+	}
+
+	return m.localNamedPath(p.thePackage), nil
+}
+
+// TODO(adam): test.
+func (m *cloudManager) RefPath(refPath string) (string, error) {
+	ref := config.GetPackageReference(refPath)
+
+	// If no reference just return original path.
+	if ref == nil {
+		return refPath, nil
+	}
+
+	packagePath, err := m.PackagePath(PackageName(ref.Package))
+	if err != nil {
+		return "", err
+	}
+
+	return path.Join(packagePath, path.Clean(ref.PathInPackage)), nil
+}
+
+// Close manager.
+func (m *cloudManager) Close() error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	m.httpClient.CloseIdleConnections()
+	return nil
+}
+
+// SyncAll syncs all given packages and removes any not in the list from the local file system.
+func (m *cloudManager) Sync(ctx context.Context, packages []config.PackageConfig) error {
+	var outErr error
+
+	// Only allow one rdk process to operate on the manager at once. This is generally safe to keep locked for an extended period of time
+	// since the config reconfiguration process is handled by a single thread.
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	newManagedPackages := make(map[PackageName]*managedPackage, len(packages))
+
+	for idx, p := range packages {
+		select {
+		case <-ctx.Done():
+			return multierr.Append(outErr, errors.New("interrupted"))
+		default:
+		}
+
+		start := time.Now()
+		m.logger.Debugf("Starting package sync [%d/%d] %s:%s", idx+1, len(packages), p.Package, p.Version)
+
+		// Package exists in known cache.
+		existing, ok := m.managedPackages[PackageName(p.Name)]
+		if ok {
+			if existing.thePackage.Package == p.Package && existing.thePackage.Version == p.Version {
+				m.logger.Debug("  Package already managed, skipping")
+
+				newManagedPackages[PackageName(p.Name)] = existing
+				delete(m.managedPackages, PackageName(p.Name))
+				continue
+			}
+			// anything left over in the m.managedPackages will be cleaned up later.
+		}
+
+		// Lookup the packages http url
+		resp, err := m.client.GetPackage(ctx, &pb.GetPackageRequest{
```

Comment:
This now needs to set the new `IncludeUrl` field in order to generate a signed URL.

Context:
API: https://github.com/viamrobotics/api/pull/186
App: https://github.com/viamrobotics/app/pull/1143

Reply:
Ack, added the field to the req